testing/jormungandr-testing-utils/src/testing/vit/vote_cast_register.rs

```diff
@@ -0,0 +1,527 @@
+        !self.is_drained()
+    }
+
+    pub fn is_drained(&self) -> bool {
+        self.register.iter().all(|x| x.is_drained())
+    }
+
+    pub fn advance_single_unsafe(&mut self) -> Result<Vec<VotesToCast>, Error> {
+        self.advance_batch_unsafe(1)
+    }
+
+    pub fn can_send_next_batch(&self, requested_batch_size: usize) -> bool {
+        self.available_to_send() > requested_batch_size
+    }
+
+    pub fn available_to_send(&self) -> usize {
+        self.register.iter().map(|x| x.available_to_send()).sum()
+    }
+
+    pub fn advance_batch_unsafe(
+        &mut self,
+        mut requested_batch_size: usize,
+    ) -> Result<Vec<VotesToCast>, Error> {
+        if !self.can_send_next_batch(requested_batch_size) {
+            return Err(Error::NoMoreRequestsToSentForWallet {
+                requested_batch_size,
+                available_to_send: self.available_to_send(),
+            });
+        }
+
+        let mut votes_to_cast = Vec::new();
+
+        for vote_plan in self.register.iter_mut().skip_while(|x| x.is_drained()) {
+            if requested_batch_size == 0 {
+                break;
+            }
+
+            let batch_size = std::cmp::min(requested_batch_size, vote_plan.available_to_send());
+            requested_batch_size -= batch_size;
+            votes_to_cast.push(vote_plan.advance_batch_unsafe(batch_size)?);
+        }
+
+        Ok(votes_to_cast)
+    }
+
+    pub fn advance_batch(
+        &mut self,
+        mut requested_batch_size: usize,
+    ) -> Result<Vec<VotesToCast>, Error> {
+        let mut votes_to_cast = Vec::new();
+        for vote_plan in self.register.iter_mut().skip_while(|x| x.is_drained()) {
+            if requested_batch_size == 0 {
+                break;
+            }
+
+            let batch_size = std::cmp::min(requested_batch_size, vote_plan.available_to_send());
+            requested_batch_size -= batch_size;
+            votes_to_cast.push(vote_plan.advance_batch(batch_size));
+        }
+        Ok(votes_to_cast)
+    }
+}
+
+#[derive(Debug)]
+pub struct VotePlanVoteCastPosition {
+    id: VotePlanId,
+    limit: u8,
+    current: u8,
+}
+
+impl VotePlanVoteCastPosition {
+    pub fn new(id: VotePlanId, limit: u8) -> Self {
+        Self {
+            id,
+            limit,
+            current: 0,
+        }
+    }
+
+    pub fn id(&self) -> VotePlanId {
+        self.id.clone()
+    }
+
+    pub fn is_drained(&self) -> bool {
+        !self.has_next()
+    }
+    pub fn has_next(&self) -> bool {
+        self.can_send_next_batch(1)
+    }
+
+    pub fn can_send_next_batch(&self, requested_batch_size: usize) -> bool {
+        self.available_to_send() >= requested_batch_size
+    }
+
+    pub fn available_to_send(&self) -> usize {
+        (self.limit - self.current).into()
+    }
+
+    pub fn advance_single_unsafe(&mut self) -> Result<VotesToCast, Error> {
+        self.advance_batch_unsafe(1)
+    }
+    pub fn advance_batch_unsafe(
```

Comment:
Nit: I think unsafe usually conveys a particular concept in rust, which is not what it means here

Reply:
renaming